### PR TITLE
Reduce unnecessary COW on Vector by make writing explicit

### PR DESCRIPTION
**Pull request by hpvb**
This PR makes operator[] on Vector const and adds a write proxy to it. From now on writes to Vectors need to happen through the .write proxy. So for instance:

```c++
Vector<int> vec;
vec.push_back(10);
std::cout << vec[0] << std::endl;
vec.write[0] = 20;
```
Failing to use the .write proxy will cause a compilation error.

In addition COWable datatypes can now embed a CowData pointer to their data. This means that String, CharString, and VMap no longer use or derive from Vector. 

`_ALWAYS_INLINE_` and `_FORCE_INLINE_` are now equivalent for debug and non-debug builds. This is a lot faster for Vector in the editor and while running tests. The reason why this difference used to exist is because force-inlined methods used to give a bad debugging experience. After extensive testing with modern compilers this is no longer the case.

Still TODO:
* Add String and CharString proxies so that cow can be reduced there also (These will not get a .write member instead we will use proxies that 'decay' into the appropriate types there)
* Find other datatypes that (ab)use Vector as their backend storage and move them to CowData also.

Given how much files this touches I'd prefer to do this in a separate PR later this week.

### CODE REVIEW

**Comment by hpvb** [comment]
This is to discuss this API change. This change as such was requested by @reduz

**Comment by hpvb** [comment]
OK, This is slowly getting somewhere now. It should build and run on all targets now (only iphone remains untested) and it has some nice speedups in bunnymark.

File: core/cowdata.h [hunk 1]
```diff
@@ -0,0 +1,332 @@
+/*************************************************************************/
+/*  cowdata.h                                                                */
```
Comment (akien-mga) [inline core/cowdata.h]: `*/` misaligned at EOL.

Comment (hpvb) [reply core/cowdata.h]: fixed

**Comment by AndreaCatania** [comment]
I've a question, but instead to do this job on making Vector acting as PoolVector why don't replace Vector on parts of code where it was not optimal and use a PoolVector?

The problem is that now doesn't make any sense have Vector and PollVector in Godot... 
What is the difference that make me choose between them?

I don't think that this change is bad, and I'm not against it, I just dislike the fact that to do it we deleted the Vector class...

**Comment by vnen** [comment]
I think the difference is that PoolVector uses a memory pool, so you don't have guarantees about memory location or layout. A Vector OTOH is always is stored in contiguous memory. At least that's how I understand it.

**Comment by AndreaCatania** [comment]
https://github.com/godotengine/godot/blob/master/core/dvector.h#L327

The memory between elements is always contiguous, at least a PoolVector is always a vector and not a list. The Pool stand for that it make a copy on write so you don't have the guaranty of "real array" location if you don't request it using read/write functions.

And now this is the same thing that happen with vector, and I don't like it.

Also I like a lot the solution of  `vector.write[]` to write the array, and just vector[] to read, I think that this is an improvement of PoolVector so I would appreciate if it was applied to PoolVector and bring back the Vector class.

**Comment by hpvb** [comment]
@AndreaCatania what do you mean by 'bring back vector' you mean return to the previous 'copy on read' behavior of Vector? It really was a broken concept before.

**Comment by AndreaCatania** [comment]
Well could be broken but now we have two class that handle arrays with the
same concept...

At least before you had the possibility to chose which one depending on the
case, but now that are the same, which one is better poolvector or just
vector?

**Comment by hpvb** [comment]
@AndreaCatania previously `Vector<T>` tried to do CoW. Basically the hope was that the compiler would pick the correct one of `operator[]` or `const operator[] const`. This did not work, what actually happened is that when sharing a copy was made on each access, be it reading OR writing.

If what you're saying is that you'd like to have a `Vector<T>` that doesn't do CoW at all then that would actually be a new thing. This PR implements the *intended* behavior of `Vector<T>`.

I don't know if @reduz thinks there's a use-case for a non-CoW datastructure. I don't really see it, with this approach there's essentially no downside to `Vector<T>` doing CoW, apart from having to think about the const-correctness of the code.

Am I actually missing a usecase here @AndreaCatania?

**Comment by AndreaCatania** [comment]
The problem is not the COW algorithm, that indeed is an improvement... I don't like the fact that now I'm forced to use vector.write[0] to write to a vector even if I know for sure that the COW will never occur in my case, and I'm bored to write and see the vector in this way.

Before, when an implementation required to share a vector I used a PoolVector and in that case use the appropriate function to read and write was not a problem for me.

In the previous implementation with COR algorithm the vector was exactly in the way I was expecting.

So my point is since we already have a PoolVector that use COW, why don't replace it in the critical part of code where is really required?

Then if you want to implement the idea of current Vector to the PoolVector, I'm fine and happy with it because for me it is an improvement

**Comment by hpvb** [comment]
@AndreaCatania so the main issue with that approach is that PoolVector is a much 'heavier' datastructure that has some mutexes and in general does a lot more work. This work helps when you have a lot of data as theoretically the way the memory pools are handles should make it faster to operate on than the contiguous 'regular' memory used by Vector.

If you have been using PoolVector only when you thought there may be sharing but have been using it for small datastructures you've probably not been writing the fastest possible code :). It seems that this whole mess needs to be a bit better documented.

So it seems that we need to do the following:
* Asses the use of Vector vs PoolVector in existing code to see if more people have been using PoolVector as 'shareable' and Vector as 'unsharable'
* Asses whether we should add a vector-like datastructure that does not do CoW at all, basically just an Array of sorts?

I really do not think it's a good idea to 'restore' Vector back to the previous 'copy on all access' implementation. Given the rather large performance delta of this patch I assume many people have been relying on it in a wrong way so this does actually fix something.

**Comment by Ranoller** [comment]
I have to say thanks very much for this commit, if it helps to the press note for 3.1 (@akien-mga ) i have to note that with this commit a time for a text parse for all entire files in a project with a gdscript plugin, counting files, file size, and files by type, lines of text files, doing a tree, etc.... was reduced in a project with 945 files, 87200KB of files and more than 100.000 lines of text (gd,tscn,tres,json,txt,etc....) in that way:

    After this commit :   +5 sec

   Before this commit:    1 sec aprox.

So.... this is a great moment to gdscipt and godot API.

I test my own bunnymark and with this commit there is only a little improvement (about 400 bunnys before reach 60fps), but do not care about this small test, a lot of API calls are now faster (x5 in some cases). Thanks a lot  @hpvb !

**Pull request merged**